src/core/qgsvectordataprovider.h

```diff
@@ -160,6 +160,21 @@ class CORE_EXPORT QgsVectorDataProvider : public QgsDataProvider, public QgsFeat
      */
     long featureCount() const override = 0;
 
+    /**
+     * Returns true if the layer contains at least one feature.
+     *
+     * \since QGIS 3.2
+     */
+    virtual bool empty() const override;
```

Comment:
Can you please implement these in QgsProcessingFeatureSource (proxying off to the underlying source to ensure any optimised overrides are used correctly?)